### ISSUE TITLE
proposal: submit intent to upload before uploading blobs

### ISSUE THREAD

**Issue by vikstrous** [docker/distribution#2174]
Right now we have a number of issues caused by the order of operations of blobs vs manifests during pushes. Blobs are uploaded first and the manifest is uploaded last. This results in:

* Not knowing how big the image the user is trying to push is -> Can't enforce storage usage quotas
* Not knowing the full list of layers ahead of time before the upload begins -> Requiring read-only mode to perform GC because otherwise a GC'd layer can be referenced by a manifest upload

If we added an extra request before an image upload starts which represents an "intent to upload" then the registry can increase the reference count on the necessary layers, tell the user which of the layers they want to push already exists, and confirm that the user is allowed to push the amount of data they intend to push. This will improve performance for pushes of images where most layers already exist and unblock the two use cases above: usage quotas and online GC.

Of course, there will always be old clients, so we can't immediately build online GC, but at least we can restrict pushes to docker 1.14.x+ clients.

**Comment by aaronlehmann** [comment]
> If we added an extra request before an image upload starts which represents an "intent to upload" then the registry can increase the reference count on the necessary layers, tell the user which of the layers they want to push already exists, and confirm that the user is allowed to push the amount of data they intend to push. This will improve performance for pushes of images where most layers already exist and unblock the two use cases above: usage quotas and online GC.

I'm not sure how this would enable online GC. What if GC ran simultaneously with the "intent to upload" endpoint? It seems to me like you need strong consistency to support online GC, but (some of) the registry's storage backends are eventually consistent.

**Comment by tonyhb** [comment]
+1 it would be great to have a list of all proposed digests to be uploaded before they begin. Additional benefits:

- We can then prevent linking to the blobstore if the digest doesn't match what we have in a digest.
- And we'll know the size etc. to carry out sanity checking when a blob upload finishes.
- We'll also be able to link blobs to any additional metadata store prior to blobs being uploaded, so if an upload is unfinished we still have blobs referenced in the metadata store
- We'll be able to detect a plugin vs manifest upload prior to uploading any blobs

**Comment by marcusmartins** [comment]
cc @nwt

**Comment by vikstrous** [comment]
@aaronlehmann Yes, sorry for online GC, we do need consistent storage, but only for the metadata, which we have in DTR.

**Comment by stevvooe** [comment]
To tell you the truth, we have avoided features that can support hard quotas, as they tend to be very anti-user. A better experience is to report and collect on quota violation. Workflow interruption is a real thing.

You can already post a manifest and get back errors about whether or not it will work. That is part of the design of the protocol. We can decorate that error set with stuff about size and quotas for the upload. Right now, it just tells you which ones are missing but this can be easily extended to provide more information. This called "pre-flighting" the manifest. We don't do it today because you can tell which blobs are present by doing head requests, but this method another possible way to do it.

As far as online GC goes, this is not at all necessary for such a thing. When a blob gets uploaded, just set a reference on newly uploaded blobs such that they can't be collected for a period of time. We could trigger these temp references by pre-flighting the manifest or other methods.

**Comment by vikstrous** [comment]
Hmm... looks like there are a lot of different reasons why we think we may want pre-flights. Let's try to separate them out:

* Regarding online GC, you are right that we can do it right now. It would require keeping track of which blobs have recently seen HEAD requests and/or been recently uploaded and picking an arbitrary maximum length of time for an upload to complete. We would still need a timeout for uploads even if we didn't have a pre-flight check, so we wouldn't be gaining much there.

* Regarding speeding up uploads for images with a lot of existing layers, I think we all agree that preflights would help. They would slightly slow down uploads with completely unique layers, but I think that since most images are based on at least one base image, this trade-off makes sense.

* Regarding separating image and plugin repos, the preflight would help fail early before uploading any of the layers, which would be a much better UX.

* Regarding hard quotas, do we know the size of layers before they are uploaded? Maybe the quota can be reasonably enforced without a pre-flight anyway? It would complicate the implementation, but what I'm interested in is if it's possible, since we can't change existing released versions of docker anyway.

I didn't cover all of the use cases, @tonyhb is any particular use case you had in mind completely impossible without pre-flights?

**Comment by stevvooe** [comment]
> Regarding speeding up uploads for images with a lot of existing layers, I think we all agree that preflights would help. They would slightly slow down uploads with completely unique layers, but I think that since most images are based on at least one base image, this trade-off makes sense.

This conclusion is dubious. We are comparing a head request per image, in parallel, with a possibly slower manifest check with deep validation. Theoretically, it may speed things up, but I am not if the UX will feel a real speed up here unless you have a lot of small layers.

> Regarding hard quotas, do we know the size of layers before they are uploaded?

If we have a manifest, we must have built all the layers, but this isn't necessarily true today.

I would recommend that you spend a little time digging into the image format. This isn't work the distribution team is going to be able to perform, so having a deep understanding on your side is going to be required to propose effective changes.
